src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionDescriptorChangeProvider.cs

```diff
@@ -1,22 +1,38 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
 using Microsoft.AspNetCore.Mvc.Infrastructure;
 using Microsoft.AspNetCore.Mvc.Razor.Internal;
-using Microsoft.AspNetCore.Razor.Evolution;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.Options;
 using Microsoft.Extensions.Primitives;
 
 namespace Microsoft.AspNetCore.Mvc.RazorPages.Internal
 {
     public class PageActionDescriptorChangeProvider : IActionDescriptorChangeProvider
     {
-        private readonly RazorProject _razorProject;
+        private readonly IFileProvider _fileProvider;
+        private readonly string _searchPattern;
 
-        public PageActionDescriptorChangeProvider(RazorProject razorProject)
+        public PageActionDescriptorChangeProvider(
+            IRazorViewEngineFileProviderAccessor fileProviderAccessor,
+            IOptions<RazorPagesOptions> razorPagesOptions)
         {
-            _razorProject = razorProject;
+            if (fileProviderAccessor == null)
+            {
+                throw new ArgumentNullException(nameof(fileProviderAccessor));
+            }
+
+            if (razorPagesOptions == null)
+            {
+                throw new ArgumentNullException(nameof(razorPagesOptions));
+            }
+
+            _fileProvider = fileProviderAccessor.FileProvider;
+            _searchPattern = razorPagesOptions.Value.RootDirectory.TrimEnd('/') +  "/*.cshtml";
         }
 
-        public IChangeToken GetChangeToken() => ((DefaultRazorProject)_razorProject).Watch("**/*.cshtml");
+        public IChangeToken GetChangeToken() => _fileProvider.Watch(_searchPattern);
```

Comment:
Do we want to watch the heirarchy of cshtml or just the cshtml at the RootDirectoy? The previous version reads like heirarchy to me but this looks like we're only checking the RootDirectoy.